compiler/src/jdk.graal.compiler.libgraal/src/jdk/graal/compiler/libgraal/LibGraalFeature.java

```diff
@@ -0,0 +1,384 @@
+                OptionDescriptor descriptor = option.getDescriptor();
+                if (descriptor.isServiceLoaded()) {
+                    GraalError.guarantee(access.isReachable(option.getClass()), "%s", option.getClass());
+                    GraalError.guarantee(access.isReachable(descriptor.getClass()), "%s", descriptor.getClass());
+
+                    String name = option.getName();
+                    OptionsParser.libgraalOptions.descriptors().put(name, descriptor);
+
+                    String module = modules.get(descriptor.getDeclaringClass().getName());
+                    if (module.contains("enterprise")) {
+                        OptionsParser.libgraalOptions.enterpriseOptions().add(name);
+                    }
+                }
+            }
+        }
+    }
+
+    private BeforeCompilationAccess beforeCompilationAccess;
+
+    /**
+     * Transformer for {@code Fields.offsets} and {@code Edges.iterationMask} which need to be
+     * recomputed to use SVM field offsets instead of HotSpot field offsets.
+     */
+    class FieldOffsetsTransformer implements FieldValueTransformer {
+        /**
+         * Map from {@link Fields} objects to a (newOffsets, newIterationMask) tuple represented as
+         * a {@link java.util.Map.Entry} value.
+         */
+        private final Map<Object, Map.Entry<long[], Long>> replacements = new IdentityHashMap<>();
+
+        final Field fieldsOffsetsField;
+        final Field edgesIterationMaskField;
+
+        FieldOffsetsTransformer() {
+            fieldsOffsetsField = lookupField(Fields.class, "offsets");
+            edgesIterationMaskField = lookupField(Edges.class, "iterationMask");
+        }
+
+        void register(BeforeAnalysisAccess access) {
+            access.registerFieldValueTransformer(fieldsOffsetsField, this);
+            access.registerFieldValueTransformer(edgesIterationMaskField, this);
+        }
+
+        @Override
+        public boolean isAvailable() {
+            return beforeCompilationAccess != null;
+        }
+
+        @Override
+        public Object transform(Object receiver, Object originalValue) {
+            Map.Entry<long[], Long> repl = getReplacement(receiver);
+            if (originalValue instanceof long[]) {
+                return repl.getKey();
+            }
+            return repl.getValue();
+        }
+
+        private Map.Entry<long[], Long> getReplacement(Object receiver) {
+            synchronized (replacements) {
+                return replacements.computeIfAbsent(receiver, this::computeReplacement);
+            }
+        }
+
+        private Map.Entry<long[], Long> computeReplacement(Object receiver) {
+            Fields fields = (Fields) receiver;
+            return fields.recomputeOffsetsAndIterationMask(beforeCompilationAccess::objectFieldOffset);
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public void beforeAnalysis(BeforeAnalysisAccess access) {
```

Comment:
Nice! It's now almost pleasant to look at.
No more reflective access to the Graal classes that we want to have in libgraal :partying_face: